app/src/main/java/com/x8bit/bitwarden/data/autofill/fido2/processor/Fido2ProviderProcessorImpl.kt

```diff
@@ -124,10 +141,118 @@ class Fido2ProviderProcessorImpl(
         cancellationSignal: CancellationSignal,
         callback: OutcomeReceiver<BeginGetCredentialResponse, GetCredentialException>,
     ) {
-        // no-op: RFU
-        callback.onError(GetCredentialUnsupportedException())
+        cancellationSignal.setOnCancelListener {
+            callback.onError(GetCredentialCancellationException())
+            scope.cancel()
+        }
+
+        // If the user is not logged in, return an error.
+        val userState = authRepository.userStateFlow.value
+        if (userState == null) {
+            callback.onError(GetCredentialUnknownException("Active user is required."))
+            return
+        }
+
+        // Return an unlock action if the current account is locked.
+        if (userState.activeAccount.isVaultUnlocked.not()) {
+            val authenticationAction = AuthenticationAction(
+                title = context.getString(R.string.unlock),
+                pendingIntent = intentManager.createFido2UnlockPendingIntent(
+                    action = UNLOCK_ACCOUNT_INTENT,
+                    requestCode = requestCode.getAndIncrement(),
+                ),
+            )
+
+            callback.onResult(
+                BeginGetCredentialResponse(
+                    authenticationActions = listOf(authenticationAction),
+                ),
+            )
+            return
+        }
+
+        // Otherwise, find all matching credentials from the current vault.
+        try {
+            runBlocking {
```

Comment:
We are OK having this block?

Is this function being called from a background thread?

Reply:
Can this method be suspending?

The the `BitwardenFido2ProviderService` would launch this in a scope?

I'm unsure the ramifications of that. @SaintPatrck might know

Reply:
Or even use `runBlocking` in the `BitwardenFido2ProviderService`, that way the caller is asserting that they know they are doing async logic on their own thread.

Reply:
We don't actually need to `block` anything here. The delegate provides `callback` so that we can offload jobs to a coroutine `scope` and invoke `callback` once those jobs are done. Invoking a `callback` delegate is what notifies the OS to continue, not completion of `processGetCredentialRequest`.